lib/common/xml.c

```diff
@@ -1367,16 +1343,180 @@ mark_child_moved(xmlNode *old_child, xmlNode *new_parent, xmlNode *new_child,
+ * the same contents.
+ *
+ * If \p new_comment has the \c pcmk__xf_skip flag set, then it is automatically
+ * considered not to match.
+ *
+ * \param[in] old_comment  Old XML child element
+ * \param[in] new_comment  New XML child element
+ *
+ * \retval \c true   if \p new_comment matches \p old_comment
+ * \retval \c false  otherwise
+ */
+static bool
+new_comment_matches(const xmlNode *old_comment, const xmlNode *new_comment)
+{
+    xml_node_private_t *nodepriv = new_comment->_private;
+
+    if ((nodepriv != NULL) && pcmk_is_set(nodepriv->flags, pcmk__xf_skip)) {
+        /* @TODO Should we also return false if old_comment has pcmk__xf_skip
+         * set? This preserves existing behavior at time of writing.
+         */
+        return false;
+    }
+    if (pcmk__xml_position(old_comment, pcmk__xf_skip)
+        != pcmk__xml_position(new_comment, pcmk__xf_skip)) {
+        return false;
+    }
+    return pcmk__xc_matches(old_comment, new_comment);
+}
+
+/*!
+ * \internal
+ * \brief Check whether a new XML child element matches an old XML child element
+ *
+ * Two elements match if they have the same name and, if \p match_ids is
+ * \c true, the same ID. (Both IDs can be \c NULL in this case.)
+ *
+ * \param[in] old_element  Old XML child element
+ * \param[in] new_element  New XML child element
+ * \param[in] match_ids    If \c true, require IDs to match (or both to be
+ *                         \c NULL)
+ *
+ * \retval \c true   if \p new_element matches \p old_element
+ * \retval \c false  otherwise
+ */
+static bool
+new_element_matches(const xmlNode *old_element, const xmlNode *new_element,
+                    bool match_ids)
+{
+    if (!pcmk__xe_is(new_element, (const char *) old_element->name)) {
+        return false;
+    }
+    return !match_ids
+           || pcmk__str_eq(pcmk__xe_id(old_element), pcmk__xe_id(new_element),
+                           pcmk__str_none);
+}
+
+/*!
+ * \internal
+ * \brief Check whether a new XML child node matches an old XML child node
+ *
+ * Node types must be the same in order to match.
+ *
+ * For comments, a match is a comment at the same position with the same
+ * content.
+ *
+ * For elements, a match is an element with the same name and, if required, the
+ * same ID. (Both IDs can be \c NULL in this case.)
+ *
+ * For other node types, there is no match.
+ *
+ * \param[in] old_child  Child of old XML
+ * \param[in] new_child  Child of new XML
+ * \param[in] match_ids  If \c true, require element IDs to match (or both to be
+ *                       \c NULL)
+ *
+ * \retval \c true   if \p new_child matches \p old_child
+ * \retval \c false  otherwise
+ */
+static bool
+new_child_matches(const xmlNode *old_child, const xmlNode *new_child,
+                  bool match_ids)
+{
+    if (old_child->type != new_child->type) {
+        return false;
+    }
+
+    switch (old_child->type) {
+        case XML_COMMENT_NODE:
```

Comment:
For the life of me, I could not tell you why we ever thought it was a good idea to include comments in these diffs. The whole point should be to pick up on changes that might matter in some way. Handling comments complicates a lot of things in change tracking for no real reason AFAICT. But, as always, "someone might be relying on Pacemaker's XML diffs to detect changes in comments"..........